login/service.go

```
@@ -162,22 +163,32 @@ func (keycloak *KeycloakOAuthProvider) Perform(ctx *app.AuthorizeLoginContext, a
 	}
 
 	// First time access, redirect to oauth provider
+	redirect := ctx.Redirect
+	referrer := ctx.RequestData.Header.Get("Referer")
+	if redirect == nil {
+		if referrer == "" {
+			jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrBadRequest("Referer Header and redirect param are both empty. At least one should be specified."))
+			return ctx.BadRequest(jerrors)
+		}
+		redirect = &referrer
+	}
 
 	// store referrer in a state reference to redirect later
 	log.Info(ctx, map[string]interface{}{
 		"referrer": referrer,
+		"redirect": redirect,
 	}, "Got Request from!")
 
 	stateID := uuid.NewV4()
-	err := keycloak.saveReferrer(ctx, stateID, referrer)
+	err := keycloak.saveReferrer(ctx, stateID, *redirect, validRedirectURL)
 	if err != nil {
 		log.Error(ctx, map[string]interface{}{
 			"state":    stateID,
 			"referrer": referrer,
+			"redirect": redirect,
 			"err":      err,
 		}, "unable to save the state")
-		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrUnauthorized("Unable to save the state. " + err.Error()))
-		return ctx.InternalServerError(jerrors)
+		return err
```

@alexeykazakov don't you want to return the error as done before ?